core/src/services/routing.js

```diff
@@ -299,6 +299,47 @@ export const handleRouteChange = async (path, component, node, config) => {
     return;
   }
   try {
+    if (
+      component &&
+      typeof component.get === 'function' &&
+      component.get().unsavedChanges &&
+      component.get().unsavedChanges.isDirty
+    ) {
+      const newUrl = window.location.href;
+      const oldUrl = component.get().unsavedChanges.persistUrl;
+
+      //pretend the url hasn't been changed
+      oldUrl && history.replaceState(window.state, '', oldUrl);
+
+      if (
```

Comment:
We should make sure the component is ready to deal with unsavedChanges feature before applying any logic like `history.replaceSate`. Therefore, I would propose to have this checks also above. My proposal would be sth like this:
```
const componentImplementsUnsavedChangesFeature = component &&
  typeof component.get === 'function' &&
  component.get().unsavedChanges &&
  component.get().unsavedChanges.isDirty &&
  typeof component.showModal === 'function' &&
  typeof component.hideModal === 'function';

if (componentImplementsUnsavedChangesFeature) {
  const newUrl = window.location.href;
  const oldUrl = component.get().unsavedChanges.persistUrl;

  //pretend the url hasn't been changed
  oldUrl && history.replaceState(window.state, '', oldUrl);
```

Reply:
As stated in another comment, this might not be necessary: `typeof component.get === 'function' &&`

Reply:
Actually, just realised we don't need so many checks, since as stated in other of my comments, here `component` is always 'AppComponent'. This means *most of the checks are always true*. So the conditions can be reduced to:
```
if (component.get().unsavedChanges.isDirty) {
  const newUrl = window.location.href;
  const oldUrl = component.get().unsavedChanges.persistUrl;

  //pretend the url hasn't been changed
  oldUrl && history.replaceState(window.state, '', oldUrl);
```

Reply:
They are not in the unit tets, that's why I wrote so many checks.